src/main/java/org/myrobotlab/client/Client.java

```diff
@@ -43,330 +46,487 @@
+            // ioe.printStackTrace();
+          }
+        }).on(Event.STATUS.name(), new Function<String>() {
+          @Override
+          public void on(String t) {
+            // System.out.println("STATUS " + t);
+          }
+        }).on(Event.HEADERS.name(), new Function<String>() {
+          @Override
+          public void on(String t) {
+            // System.out.println("HEADERS " + t);
+          }
+        }).on(Event.MESSAGE_BYTES.name(), new Function<String>() {
+          @Override
+          public void on(String msg) {
+            // System.out.println("MESSAGE_BYTES " + t);
+            System.out.println(msg);
+          }
+        }).on(Event.OPEN.name(), new Function<String>() {
+          @Override
+          public void on(String t) {
+            // System.out.println("OPEN " + t);
+          }
+        }).open(request.build());
+      }
 
-      return new StringReader(data);
-      // return null;
+    } catch (Exception e) {
+      // the exception is propagated to socket event
+      // handler too - no need to double print
+      // System.out.println(e.getMessage());
     }
   }
 
-  Set<RemoteMessageHandler> handlers = new HashSet<>();
+  public String getUrl() {
+    return String.format("%s/api/%s?id=%s", base, apiType, clientId);
+  }
 
-  public class AutoConnector implements Runnable {
-    transient Thread worker;
-    boolean isRunning = false;
-    long interval = 1000;
+  @Override
+  public Reader decode(Event e, String s) {
+    // System.out.println("decode " + s);
+    if ("X".equals(s)) {
+      return null;
+    }
+    if ("OPEN".equals(s)) {
+      System.out.println("connected to " + getUrl());
+      return null;
+    }
 
-    public void run() {
-      isRunning = true;
-      while (isRunning) {
-        try {
-          Thread.sleep(interval);
-        } catch (Exception e) {
-          isRunning = false;
-        }
-        for (String url : endpoints.keySet()) {
-          try {
-            if (endpoints.get(url).socket == null) {
-              connect(url);
-            }
-          } catch (Exception e) {
-            log.error("could not connect to {}", url, e);
-          }
-        }
-      }
-      isRunning = false;
-      worker = null;
+    if ("CLOSE".equals(s)) {
+      System.out.println("disconnected from " + getUrl());
+      return null;
     }
 
-    synchronized public void start() {
-      if (worker == null) {
-        worker = new Thread(this, "auto-connector");
-        worker.start();
-      }
+    // System.out.println(s);
+    Message msg = gson.fromJson(s, Message.class);
+
+    if ("describe".equals(msg.method)) {
+      serverHelloRequest = gson.fromJson(msg.data[1].toString(), DescribeQuery.class);
+      promptId = serverHelloRequest.id;
+      System.out.println("attaching to id " + promptId);
     }
 
-    synchronized public void stop() {
-      isRunning = false;
+    if (msg.data == null || msg.data[0] == null) {
+      System.out.println("null");
```

Comment:
we should probably clean  up stdout .. i suspect it's only here for debugging?

Reply:
Client is not used by anything - its standalone, part of it was the interest of a single file client with no dependencies ...
So, no slf4j/log4j etc...
But, I've kindof waffled a bit, because it currently uses picocli :P